### PR TITLE
reset solver when copying parameters

### PR DESCRIPTION
**Pull request by kylebaron**
Issue #745

- reset solver any time parameters are copied from the data
- see the issue; there were solver overshoot issues when time varying covariates changed parameter values; this is fixed by re-initializing solver whenever we copy parameters

### CODE REVIEW

**Review by yizhang-yiz — COMMENTED** [review]
Can we just call `prob->lsoda_init` inside `copy_parameters` or there are other scenarios when we need copy param but don't want to reinitialize `prob`? 

When `istate=1` is set iby default lsoda will redo time step adaptation, this involves overhead but I guess it's hard for us to know what types of covariant change require re-init. What you've tried in the other thread with small hmax essentially forces solver to climb a possible discontinuity, which could be time consuming and inaccurate.

**Comment by kylebaron** [comment]
@yizhang-yiz Yeah, I was trying to avoid doing this if possible b/c it does slow things down.  There has always been some inaccuracy around this, but honestly it seems to be worse since I switched to that C++ implementation of dlsoda.  It's hard to say but seems that way.  I'm not sure if there is anything we could look into in the solver or some way we could do what we need for the reset without giving up time step adaptation.  Is this worth looking into?

**Review by yizhang-yiz — APPROVED** [review]

**Comment by yizhang-yiz** [comment]
>  Is this worth looking into?

In general yes but that'll depend on how much the two solvers diverges and the degree of slowdown. Since I can't think of any way to detect parameters that would cause unnecessary reinitialization, the current design makes sense to me.

File: src/dataobject.cpp [hunk 1]
```diff
@@ -186,10 +186,10 @@ void dataobject::idata_row() {
 
 void dataobject::copy_parameters(int this_row, odeproblem* prob) {
   size_t n = par_from.size();
-  for(size_t i=0; i < n; ++i) {
+  for(size_t i = 0; i < n; ++i) {
     prob->param(par_to[i],Data(this_row,par_from[i]));
   }
-  prob->lsoda_init();
+  if(n > 0) prob->lsoda_init();
```
Comment (yizhang-yiz) [inline src/dataobject.cpp:192]: Doesn't `any_copy` control this? This is redundant if `copy_parameters` is called only when `any_copy==true`.

Comment (kylebaron) [reply src/dataobject.cpp:192]: Yeah; thanks.

**Review by yizhang-yiz — APPROVED** [review]

**Comment by kylebaron** [comment]
@yizhang-yiz what do you think of this?  I've been benchmarking this a bit and this seems to take the edge off some of the slow down when parameters aren't changing from record to record (which I think is the main scenario where we could change things; if covariates are time-varying).

**Comment by yizhang-yiz** [comment]
It makes sense. What does the benchmark say about the performance difference?

**Comment by kylebaron** [comment]
# Before we changed anything


run a - `data1` is sparse input data, with constant covariate
run b - `data2` is sparse input data, without covariate
run c - `data3` is rich input with constant covariate
run d - `data4` is rich input data with no covariate

All runs have same duration and same number of output records.  There is a 
small penalty for rich input data (maybe more input processing).


```r
Unit: milliseconds
                    expr       min        lq      mean    median
 a <- mrgsim(mod, data1)  7.901581  8.800033 10.650774  9.310863
 b <- mrgsim(mod, data2)  7.968120  8.849370  9.741316  9.349273
 c <- mrgsim(mod, data3) 10.487445 12.019940 16.110642 12.752516
 d <- mrgsim(mod, data4) 10.191776 11.817085 14.856134 12.435446
```


# What we had yesterday

I made this to run slowly; simulated 3 days with every 15 minutes input record;
when there is a covariate with rich input(run c `data2`) we have to stop the 
solver every record and the time increases.  With sparse input, it doesn't make 
that big of a difference (it's like a handful of extra stops ... hard to detect).


```r
Unit: milliseconds
                    expr       min        lq      mean    median 
 a <- mrgsim(mod, data1)  7.741893  8.533279  9.791214  9.007484 
 b <- mrgsim(mod, data2)  7.795728  8.485864  9.774564  8.957434 
 c <- mrgsim(mod, data3) 35.744272 38.427414 44.985757 39.986618 
 d <- mrgsim(mod, data4) 10.219277 11.148762 13.466045 11.953646 
```



# With most recent update

With the most recent proposal; this works much faster because it doesn't restart
if the covariate value doesn't change.  If we had rich input with time-varying
covariate, you'd be back to the 44 ms run time.  But that seems to be the least
likely thing to happen.

But the penalty is that we have to do the extra work of comparing parameters 
comparison every time we update.


```r
Unit: milliseconds
                    expr       min        lq     mean    median 
 a <- mrgsim(mod, data1)  8.012418  8.991715 10.03263  9.538467 
 b <- mrgsim(mod, data2)  7.891747  8.980107 10.62298  9.505040 
 c <- mrgsim(mod, data3) 10.676537 12.227749 15.14205 12.990485 
 d <- mrgsim(mod, data4) 10.626696 11.882866 14.19894 12.798136 
```

**Comment by kylebaron** [comment]
Thanks again for your help, @yizhang-yiz ; I feel good about this.

**Pull request merged**